crates/bevy_ecs/src/schedule/graph/graph_map.rs

```diff
@@ -0,0 +1,480 @@
+        self.nodes.contains_key(&n)
+    }
+
+    /// Add an edge connecting `a` and `b` to the graph.
+    /// For a directed graph, the edge is directed from `a` to `b`.
+    ///
+    /// Inserts nodes `a` and/or `b` if they aren't already part of the graph.
+    pub(crate) fn add_edge(&mut self, a: NodeId, b: NodeId) {
+        if self.edges.insert(Self::edge_key(a, b)) {
+            // insert in the adjacency list if it's a new edge
+            self.nodes
+                .entry(a)
+                .or_insert_with(|| Vec::with_capacity(1))
+                .push(CompactNodeIdAndDirection::store(b, Outgoing));
+            if a != b {
+                // self loops don't have the Incoming entry
+                self.nodes
+                    .entry(b)
+                    .or_insert_with(|| Vec::with_capacity(1))
+                    .push(CompactNodeIdAndDirection::store(a, Incoming));
+            }
+        }
+    }
+
+    /// Remove edge relation from a to b
+    ///
+    /// Return `true` if it did exist.
+    fn remove_single_edge(&mut self, a: &NodeId, b: &NodeId, dir: Direction) -> bool {
+        let Some(sus) = self.nodes.get_mut(a) else {
+            return false;
+        };
+
+        let Some(index) = sus
+            .iter()
+            .copied()
+            .map(CompactNodeIdAndDirection::load)
+            .position(|elt| (DIRECTED && elt == (*b, dir)) || (!DIRECTED && elt.0 == *b))
+        else {
+            return false;
+        };
+
+        sus.swap_remove(index);
+        true
+    }
+
+    /// Remove edge from `a` to `b` from the graph.
+    ///
+    /// Return `false` if the edge didn't exist.
+    pub(crate) fn remove_edge(&mut self, a: NodeId, b: NodeId) -> bool {
+        let exist1 = self.remove_single_edge(&a, &b, Outgoing);
+        let exist2 = if a != b {
+            self.remove_single_edge(&b, &a, Incoming)
+        } else {
+            exist1
+        };
+        let weight = self.edges.remove(&Self::edge_key(a, b));
+        debug_assert!(exist1 == exist2 && exist1 == weight);
+        weight
+    }
+
+    /// Return `true` if the edge connecting `a` with `b` is contained in the graph.
+    pub fn contains_edge(&self, a: NodeId, b: NodeId) -> bool {
+        self.edges.contains(&Self::edge_key(a, b))
+    }
+
+    /// Return an iterator over the nodes of the graph.
+    pub fn nodes(
+        &self,
+    ) -> impl DoubleEndedIterator<Item = NodeId> + ExactSizeIterator<Item = NodeId> + '_ {
+        self.nodes.keys().copied()
+    }
+
+    /// Return an iterator of all nodes with an edge starting from `a`.
+    pub fn neighbors(&self, a: NodeId) -> impl DoubleEndedIterator<Item = NodeId> + '_ {
+        let iter = match self.nodes.get(&a) {
+            Some(neigh) => neigh.iter(),
+            None => [].iter(),
+        };
+
+        iter.copied()
+            .map(CompactNodeIdAndDirection::load)
+            .filter_map(|(n, dir)| (!DIRECTED || dir == Outgoing).then_some(n))
```

Comment:
Why do we only use Outgoing edges for directed graphs? aren't neighbors also incoming edges?

Reply:
This is a hangover from `petgraph`, but I believe the rationale is that a neighbour is something you can _go to_ from your current node. So in `a <-> b`, both `a` and `b` are neighbours. Whereas, in `a --> b`, `b` is a neighbour of `a`, but not the other way around. I could change this, but I wanted to keep the API here as-close-as possible to `petgraph` to avoid subtle translation errors for contributors moving to the new type.